hnix-store-core/src/System/Nix/Internal/Nar.hs

```diff
@@ -0,0 +1,243 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeApplications    #-}
+
+{-|
+Description : Effects and functions for interacting with Nar files.
+|-}
+module System.Nix.Internal.Nar where
+
+import           Control.Applicative
+import           Control.Monad              (replicateM, replicateM_, (<=<))
+import qualified Data.Binary                as B
+import qualified Data.Binary.Get            as B
+import qualified Data.Binary.Put            as B
+import           Data.Bool                  (bool)
+import qualified Data.ByteString            as BS
+import qualified Data.ByteString.Char8      as BSC
+import qualified Data.ByteString.Lazy       as BSL
+import           Data.Foldable              (forM_)
+import qualified Data.Map                   as Map
+import           Data.Maybe                 (fromMaybe)
+import           Data.Monoid                ((<>))
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           Data.Traversable           (forM)
+import           GHC.Int                    (Int64)
+import           System.Directory
+import           System.FilePath
+
+-- | Copied from @RawFilePath@ in the @unix@ package, duplicated here
+-- to avoid the dependency.
+type RawFilePath = BS.ByteString
+
+data NarEffects (m :: * -> *) = NarEffects {
+    narReadFile   :: RawFilePath -> m BSL.ByteString
+  , narWriteFile  :: RawFilePath -> IsExecutable -> BSL.ByteString -> m ()
+  , narListDir    :: RawFilePath -> m [FilePathPart]
+  , narCreateDir  :: RawFilePath -> m ()
+  , narCreateLink :: RawFilePath -> RawFilePath -> m ()
+  , narIsExec     :: RawFilePath -> m IsExecutable
+  , narIsDir      :: RawFilePath -> m Bool
+  , narIsSymLink  :: RawFilePath -> m Bool
+  , narFileSize   :: RawFilePath -> m Int64
+  , narReadLink   :: RawFilePath -> m RawFilePath
+}
+
+
+-- Directly taken from Eelco thesis
```

Comment:
Do we know what this comment was referring to? The whole module, Nar effects? Probably not the Nar newtype.